scripts/mixin/thaumcraft.zs

```diff
@@ -155,30 +159,95 @@ zenClass MixinScanSky {
     }
 }
 
-/*
-#mixin {targets: "thaumcraft.common.lib.events.ToolEvents"}
+#mixin {targets: "thaumcraft.common.lib.events.ToolEvents", priority: 1100}
 zenClass MixinToolEvents {
     #mixin Static
-    #mixin Overwrite
-    function doRefining(event as HarvestDropsEvent, heldItem as ItemStack) as void {
+    #mixin Inject {method: "doRefining", at: {value: "HEAD"}, cancellable: true}
+    function doBuffedRefining(event as HarvestDropsEvent, heldItem as ItemStack, ci as CallbackInfo) as void {
+        ci.cancel();
+
         val level = EnumInfusionEnchantment.getInfusionEnchantmentLevel(heldItem, EnumInfusionEnchantment.REFINING);
-        val chance = 0.5f * level;
-        var b = false;
-
-        for i, is in event.drops {        
-            val cluster = Utils.findSpecialMiningResult(is, chance, event.world.rand);
-            if (!is.isItemEqual(cluster)) {
-                if (level >= 3 && event.world.rand.nextFloat() > 1.0f / (level - 1)) {
-                    cluster.count = cluster.count * 2;
+        val chance_percent = 40 + 20 * (level - 1) + 10 * event.fortuneLevel;
+
+        val lucky_number = event.world.rand.nextInt(100);
+        if (lucky_number >= chance_percent) { // if chance < 1, refining may not trigger at all
+            return;
+        }
+
+        var dropAmount = chance_percent / 100;
+        if (lucky_number < chance_percent % 100) {
+            dropAmount += 1;
+        }
+
+        var outputMultiplier = 1;
+        for oreID in OreDictionary.getOreIDs(ItemStack(event.getState().getBlock())) {
+            val newName = OreDictionary.getOreName(oreID);
+            if (isNull(newName)) continue;
+            
+            if (newName.startsWith("oreNether") || newName.startsWith("oreEnd")) {
+                outputMultiplier = 2;
+                break;
+            }
+        }
+
+        var newDrops = [] as ItemStack[];
```

Comment:
Using list instead of Array should highly increase performance:
```ts
var newDrops as [ItemStack] = [];
```
In this case, when you add elements, it wont every time create new array with new length.